utils.pwsh/Setup-Target.ps1

```diff
@@ -77,29 +77,15 @@ function Find-VisualStudio {
             Find-VisualStudio
     #>
 
-    $VisualStudioData = Get-CimInstance MSFT_VSInstance
-
-    # Prefer VS versions in this order:
-    # 1. VS2022 Release (stable)
-    # 2. VS2022 Preview
-    # 3. VS2019 Release
-    [string[]]$SupportedVSVersions =
-        "VisualStudio.17.Release",
-        "VisualStudio.17.Preview",
-        "VisualStudio.16.Release"
-    $NumSupportedVSVersions = $SupportedVSVersions.length
-
-    if ( $VisualStudioData.GetType() -eq [object[]] ) {
-        for ( $i = 0; $i -lt $NumSupportedVSVersions; $i++ ) {
-            $VisualStudioDataTemp = ($VisualStudioData | Where-Object {$_.ChannelId -eq $SupportedVSVersions[$i]} | Sort-Object -Property Version)[0]
-            if ( $VisualStudioDataTemp ) {
-                break;
-            }
-        }
-        $VisualStudioData = $VisualStudioDataTemp
+   $VSSetupData = Get-InstalledModule VSSetup -ErrorAction SilentlyContinue
+
+    if ( $VSSetupData -eq $null ) {
+        Install-Module VSSetup -Scope CurrentUser
     }
 
-    if ( ! ( $VisualStudioData ) -or ( $VisualStudioData.Version -lt 16 ) ) {
+    $VisualStudioData = Get-VSSetupInstance | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest
```

Comment:
If a user has VS2022 Stable and Preview both installed, which will it select? The system I would normally check this on is currently not up. The docs say, "By default, only releases are shown." unless you specify the `-Prerelease` flag, but a confirmation would be nice. If this does indeed only list stable releases, then this would be a slight behavior change, as we previously selected the first available from 2022 Release, 2022 Preview, 2019 Release. This should be fine on CI where only Release is available, though it could break on systems that only have VS2022 Preview installed.

Reply:
By default it should only pick up normal releases sorted by version and installation date. Prerelease versions are only included when the `-prerelease` or `-all` switch is added. This mirrors the behaviour of `vswhere` (which is on purpose iirc).

Reply:
Hmm. Our old behavior checked for VS2022 Release, then VS2022 Preview, then VS2019 Release. If we want to mimic that behavior, we might have to call this multiple times for each specific version. Otherwise, we abandon that behavior and accept that it will fail on systems that only have VS2022 Preview installed.

Reply:
I think if we add `-prerelease` the choice would still be based on version number. So If VS2022 17.1 and VS2022 17.1 Preview are installed (not that this is a probable situation), I guess the release version is "more recent". VS2022 17.2 Preview would _be_ more recent however. So it would always chose whatever has the higher version number.

To that end we could just add that one to also include preview versions.

Reply:
If 17.6 Release and 17.7 Preview were both installed and `-Prerelease -Latest` was specified, I imagine it would pick 17.7 Preview. The real question is what if the user has 17.7 Release and 17.7 Preview installed.

That said, the previous selection process didn't even bother taking this into account, it simply preferred:
1. VS2022 Release (any version)
2. VS2022 Preview (any version)
3. VS2019 Release (any version)

If you had VS2022 17.3 Release and VS2022 17.7 Preview installed, it would select VS23022 17.3 Release. This was, presumably, to allow devs to have the Release version that we targeted installed while also having some other Preview version installed if they needed it for other dev purposes outside of OBS.

Will think on it a bit.

Reply:
Release and Preview would probably differ in their patch or build versions, so that would be taken into account. So if Release is build 6521 and Preview is 6532, then Preview "wins".

I would expect this to be a very simple version number comparison, only one variant includes the version numbers of preview builds and the other doesn't.

Reply:
Tested with 17.5 Release and 17.7 Preview.
```pwsh
PS C:\> Get-VSSetupInstance | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest

InstanceId          : 142e8861
DisplayName         : Visual Studio Community 2022
InstallationVersion : 17.5.33627.172
InstallationPath    : C:\Program Files\Microsoft Visual Studio\2022\Community
InstallDate         : 11/1/2022 1:38:25 PM


PS C:\> Get-VSSetupInstance -Prerelease | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest

InstanceId          : 3d2ad485
DisplayName         : Visual Studio Community 2022
InstallationVersion : 17.7.33711.374
InstallationPath    : C:\Program Files\Microsoft Visual Studio\2022\Preview
InstallDate         : 5/20/2023 1:40:40 PM
```

Reply:
That at least worked as expected. It is still just a little unfortunate that it differs from current behavior. The only way I can think of to get the current behavior is to do multiple calls.

```pwsh
$VSInstance = Get-VSSetupInstance | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest
// some null/empty/false check
$VSInstance = Get-VSSetupInstance -Prerelease | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest
```

Reply:
I'd argue that the old behaviour was actually broken insofar as even a more recent VS 2022 preview version would not have been picked up.

The only scenario that code was made for was for the transitional period for when there was only a preview version of VS 2022 but a release was about to happen and we wanted the jobs not to suddenly break on CI.

Reply:
> The only scenario that code was made for was for the transitional period for when there was only a preview version of VS 2022 but a release was about to happen and we wanted the jobs not to suddenly break on CI.

That's not what it was for. GitHub Actions images never have VS Preview builds, only Release builds. If I recall correctly, it was intended to provide build stability for _local_ builds where a dev might have multiple versions of VS installed for _some_ reason (work, other dev projects, etc.), but still wanted to be able to build obs-deps.

Reply:
Fixed via adding explicit prerelease switch to main build script toggling inclusion of Preview builds in version checks.